docs/proposals/workflow.md

```diff
@@ -0,0 +1,281 @@
+    // Steps contains the vertices of the workflow graph. The key of the map is a string
+    // to uniquely identify the step. Steps order is defined by their dependencies.
+    Steps map[string]WorkflowStep `json:"steps,omitempty"`
+}
+```
+
+* `spec.steps`: is a map of `WorkflowStep`s. _Key_ of the map is a string which identifies the step.
+
+
+### `WorkflowStep`
+
+The `WorkflowStep` resource acts as a [union](https://en.wikipedia.org/wiki/Tagged_union) of `JobSpec` and `ObjectReference`.
+
+```go
+// WorkflowStep contains necessary information to identifiy the node of the workflow graph
+type WorkflowStep struct {
+    // JobTemplate contains the job specificaton that should be run in this Workflow.
+    // Only one between externalRef and jobTemplate can be set.
+    JobTemplate JobSpec `json:"jobTemplate,omitempty"`
+
+    // External contains a reference to another schedulable resource.
+    // Only one between ExternalRef and JobTemplate can be set.
+    ExternalRef api.ObjectReference `json:"externalRef,omitempty"`
+
+    // Dependecies represent dependecies of the current workflow step
+    Dependencies ObjectDependencies `json:"dependencies,omitempty"`
+}
+```
+
+* `workflowStep.jobSpec` contains the specification of the job to be executed.
+* `workflowStep.externalRef` contains a reference to external resources (for example another `Workflow`).
+*
+
+```go
+type ObjectDependencies struct {
+    // DependeciesRef is a slice of unique identifier of the step (key of the spec.steps map)
+    DependencyIDs []string `json:"dependencyIDs,omitempty"`
+    ControllerRef *ObjectReference `json:"controllerRef,omitempty"`
+    //...
+}
+```
+
+* `dependencies.dependencyIDs`: is a slice with a list of _step_ that must run to completion.
+* `dependencies.controllerRef`: will contain the controller for the current `WorkflowStep`. As a first
+
+
+This approach permits to implement other kinds of controller, for example data availability
+or other external event. In a first implementation `dependencies.controllerRef` will implement only
+the logic to check all dependencies ran to completion: since at the beginning only `Workflow` and `Job`
+can be composed the only thing needed to implement is the ability to check wether a `Job` or
+a `Workflow` runs to completion.
+Our understanding is that detecting the type of object and an approach similar to what
+is implemented in `pkg/client/unversioned/conditions.go` and  `pkg/kubectl/scale.go` for _desiredReplicas_ can
+be used to to detect if a _step_ must be started.
+
+
+### `WorkflowStatus`
+
+```go
+// WorkflowStatus contains the current status of the Workflow
+type WorkflowStatus struct {
+    // Conditions represent the latest available observations of an object's current state.
+    Conditions []WorkflowCondition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+
+    // Statuses represent status of different steps
+    Statuses map[string]WorkflowStepStatus `json:statuses`
```

Comment:
Does it make sense to create the `WorkflowStepStatus` object if it only holds the reference for the object? I'd place the reference here and that should be it, unless you're planning to extend WSS, but I don't see the need for that in the near feature since the Object will contain enough of a information about itself.